<html>

<head>
<meta http-equiv=Content-Type content="text/html; charset=windows-1252">
<meta name=Generator content="Microsoft Word 15 (filtered)">
<style>
<!--
 /* Font Definitions */
 @font-face
	{font-family:"Cambria Math";
	panose-1:2 4 5 3 5 4 6 3 2 4;}
@font-face
	{font-family:Calibri;
	panose-1:2 15 5 2 2 2 4 3 2 4;}
 /* Style Definitions */
 p.MsoNormal, li.MsoNormal, div.MsoNormal
	{margin-top:0in;
	margin-right:0in;
	margin-bottom:8.0pt;
	margin-left:0in;
	line-height:107%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
a:link, span.MsoHyperlink
	{color:blue;
	text-decoration:underline;}
a:visited, span.MsoHyperlinkFollowed
	{color:#954F72;
	text-decoration:underline;}
p
	{margin-right:0in;
	margin-left:0in;
	font-size:12.0pt;
	font-family:"Times New Roman",serif;}
.MsoChpDefault
	{font-family:"Calibri",sans-serif;}
.MsoPapDefault
	{margin-bottom:8.0pt;
	line-height:107%;}
@page WordSection1
	{size:8.5in 11.0in;
	margin:1.0in 1.0in 1.0in 1.0in;}
div.WordSection1
	{page:WordSection1;}
-->
</style>

</head>

<body lang=EN-US link=blue vlink="#954F72">

	<div class=WordSection1>

		<p class=MsoNormal>
			<span style='font-size:12.0pt;line-height:107%'>
				RepeatsPlus
				program finds maximal repeats in input sequences.
			</span>
		</p>

		<p class=MsoNormal><i><span style='color:#4472C4'>Option Fasta file:</span></i></p>

		<p class=MsoNormal>Fasta file specifies input file that is expected to be in Fasta format. </p>

		<p class=MsoNormal><i><span style='color:#4472C4'>Option Sequence type:</span></i></p>

		<p class=MsoNormal>
			Sequence type specifies the input sequence type (DNA, RNA or
			Protein).
		</p>

		<p class=MsoNormal><i><span style='color:#4472C4'>Option Minimal repeat length:</span></i></p>

		<p class=MsoNormal>
			Minimal length specifies the minimal length of the repeats
			that are extracted from input file.
		</p>

		<p class=MsoNormal><i><span style='color:#4472C4'>Option Repeat type:</span></i></p>

		<p class=MsoNormal>RepeatType specifies the type of repeats that are extracted. Possible types are direct non-complementary, indirect non-complementary, direct complementary and indirect complementary. </p>

		<p class=MsoNormal><i><span style='color:#4472C4'>Option MaxGap:</span></i></p>

		<p class=MsoNormal>
			MaxGap specifies the maximal gap between two components that make the repeat. If
			maxgap is zero, tandem repeats are found. If not specified all repeats are
			found (not depending on their gap).
		</p>

		<p class=MsoNormal><i><span style='color:#4472C4'>Option Ambigious letters:</span></i></p>

		<p class=MsoNormal>
			Ambigious letters option specifies that if there are
			ambigious letters in input sequence, repeats are found considering all possibilities
			for ambigious letters.
		</p>

		<p class=MsoNormal><i><span style='color:#4472C4'>Option motif:</span></i></p>

		<p class=MsoNormal>Motif specifies a mask that is used to find all repeats that satisfy given motif mask. In defining motif mask next rules are used:</p>
		<ul>
			<li>string of characters drawn from the alphabet denotes a sequence of the corresponding characters; <br>
			<li>character '.' matches any character from alphabet;<br>
			<li>characters in brackets denote any one caracter from brackets;<br>
			<li>characters in brackets starting with ^ denote any one character from alphabet that is not in the brackets.<br>
		</ul>

		<p>If option mismatch is checked then found repeats may have a mismatch. In this case statistical filtering is not provided.</p>

		<p class=MsoNormal>
			<i>
				<span style='color:#4472C4;background:white'>
					Option P
					value:
				</span>
			</i>
		</p>

		<p class=MsoNormal>
			P value specifies number that is used as p value in case of
			statistic filtering.
		</p>

		<p class=MsoNormal>
			<i>
				<span style='color:#4472C4'>
					Option Custom complement
					mapping file:
				</span>
			</i>
		</p>

		<p class=MsoNormal>
			A custom complement mapping file can be used. Specifies the
			file with the letters and their complements . The file should be a text file
			with an even number of letters. The first two letters form a pair with the
			first letter having the second letter as a complement, then the second two
			letters form another pair, etc.
		</p>

		<p class=MsoNormal><i><span style='color:#4472C4'>Option Protein wildcards:</span></i></p>

		<p class=MsoNormal>
			Different protein wildcards can be used as: Aliphatic,
			Sulphur, Tiny…
		</p>

		<p class=MsoNormal>
			<i>
				<span style='color:#4472C4'>
					Option Treat all sequence in
					Fasta file as single sequence:
				</span>
			</i>
		</p>

		<p class=MsoNormal>
			If input file contains multiple Fasta files they can be
			treated as a single input file. If this option is not chosen and input file
			contains multiple Fasta files they will be treated one by one.
		</p>

		<p class=MsoNormal><i><span style='color:#4472C4'>Option Exclude letter N segments:</span></i></p>

		<p class=MsoNormal>
			The letter N in genomic sequences is used to specify an
			unknown nucleic acid. Large islands of N's typically lead to an explosion of
			found pairs. This option instructs program to skip over any island of&nbsp;<i>number</i>&nbsp;or
			more consecutive N's if <i>number</i> is specified. Excludes every
			consecutive&nbsp;<i>number </i>or more letters N from repeat searching.  If  0
			is specified than all N will be excluded.
		</p>

		<p class=MsoNormal><i><span style='color:#4472C4'>Option Output file name:</span></i></p>

		<p class=MsoNormal>Specifies name of input file.</p>

		<p class=MsoNormal>
			<i>
				<span style='color:#4472C4'>
					Option ODBC connection
					string:
				</span>
			</i>
		</p>

		<p class=MsoNormal>
			Specifies ODBC connection string used to connect to
			database. <span style='color:black'>
				Results can be written directly either to
				DB2 or to other relational database with an ODBC driver installed. A sample
				database schema can be found in&nbsp;dbschema.txt.
			</span>
		</p>

	</div>

</body>

</html>
