<!-- saved from url=(0022)http://internet.e-mail -->
<html>
<head>
    <title>StatRepeats</title>
</head>
<body>
<h1>StatRepeats</h1>
<p> StatRepeats is a tool  for finding maximal repeats in an input (nucleotide or protein) sequence. It can extract four types of repeats: direct non-complementary, direct complementary, inverse non-complementary and inverse complementary repeats. It can use statistical estimation to extract only statistically significant repeats, or all repeats. It also can filter the set of found repeats based on the specified p-value. Results can be redirected to the standard output, to files in a few different formats, or directly into an ODBC-compatible database.</p>
<h2>Table of Contents</h2>
	<ul>
	<li><a href="#usage">Usage</a></li>
	<li><a href="#Optionalarguments">Optional arguments</a></li>	
	<ul>
		<li><a href="#fileoutput">File output options</a></li>	
		<li><a href="#seqtypes">Repeat types options</a></li>
		<li><a href="#estimation">Probability estimation</a></li>
		<li><a href="#alphabet">Options related to input sequence characteristics</a></li>		
		<li><a href="#dboutput">Database output options</a></li>			

	</ul>
	<li><a href="#download">Download</a></li>	
	<li><a href="#setup">Setup Instructions</a></li>
	<ul>
		<li><a href="#windows">Windows</a></li>
		<li><a href="#linux">Linux</a></li>
	</ul>
	</ul>

<h2 id="usage">Usage</h2>
<b>StatRepeats</b>  &lt <i>input file name</i> &gt   &lt <i>minimal length</i> &gt [optional arguments] <br>
where 
<ul>
<li><i>Input file name</i> and <i>minimal length</i> are mandatory arguments.<br>
<li><i>Input file name</i> is the name of the FASTA file to be processed. The file can contain multiple FASTA sequences, which are by default processed one by one.<br>
<li><i>Minimal length</i> specifies the minimal length of the extracted repeats. If the specified length is to low for statistical filtering to work correctly the program will suggest a minimal length. If zero is specified as the minimal length StatRepeats will automatically choose the lowest possible value for which the probabability estimation used for statistical filtering is valid.
</ul>
StatRepeats by default can handle collection of multiple fasta sequences in one file. Every sequence is processed separately; program options are applied on 
all fasta sequences. Calculation was done as series of calculation for each sequence. Output for next calculation is append at the end of previous one.
<p>Example 1: <a href="./examples/example_1/example_1.txt">calculate  (dn) repeats with minimal length 10</a>&nbsp;&nbsp;<a href="./examples/example_1/">Results</a>
<br>
Note: In this and all further examples, it is assumed that the StatRepeats program is in the same directory as the input data
</p>

<h2  id="Optionalarguments">Optional arguments</h2>

<h3 id="fileoutput">File output options</h3>
<p>By default StatRepeats will print all results to the standard output.</p>
<p><b>General output format: </b>[&lt<i>prefix</i>&gt],&lt<i>LP start position</i>&gt, &lt<i>LP end position</i>&gt, 
&lt<i>RP start position</i>&gt, &lt<i>RP end position</i>&gt, &lt<i>sequence length</i>&gt, 
&lt<i>LP</i>&gt, &lt<i>RP</i>&gt <br>
where LP and RP are the left and right parts of repeat sequence. 
<i>Prefix</i> in result is taken from fasta header in input file. If fasta header is not present prefix is null (i.e. not exists).
Both start and end position of left and right parts of repeat sequence are included in order to make furher processing more efficient.
<p><br></p>
<h4>Options related to file output</h4>
<p></p>
<TABLE CELLPADDING="0" CELLSPACING="5">
<TR ALIGN="LEFT">
<TD style="padding: 0 15px; vertical-align:top;"><a><i>-o[utput] filename</i> </a></TD>
<TD style="padding: 0 15px;"><a> Write results to a specified file. Default for output is writing results on console. When -output option is specified, other  output options are ignored </a></TD>
<TR ALIGN="LEFT">
</TR>
<TD style="padding: 0 15px; vertical-align:top;"><a><i>-loa[d] name</i></a></TD>
<TD style="padding: 0 15px;"><a>Produce three output files. Sequence name is written in file <i>name.id</i>, statistics are written in file <i>name.stat</i>, and results are written to a file <i>name.load.</i> 
If <i>-load name</i> option is combined with multiple fasta sequences in one file, then output for all sequences is placed in three files name.id, name.stat and name.load</a></TD>
</TR>
<TR ALIGN="LEFT">
</TR>
<TD  style="padding: 0 15px; width:250px; vertical-align:top;"><a><i>-sp[lit]</i></a></TD>
<TD  style="padding: 0 15px;"><a>Produce three output files. Sequence name is written in file named sequencename.id, statistics are written in file named sequencename.stat, and results are written in file named sequencename.load.
Input file must have fasta header. In opposite case, generated files with results have no prefix and start with "." (.id, .load, .stat). 
If <i>-split</i> option is combined with multiple fasta sequences in one file, then output produce, for each sequence in input file, three files with sufixes .id, .stat and .load and prefix equal with id from fasta header</a></TD></a></TD>
</TR>
<TR ALIGN="LEFT">
<TD style="padding: 0 15px; vertical-align:top;"><a> <i>-pri[ntinstanes] &lt;true | false&gt;</i></a></TD>
<TD style="padding: 0 15px;"><a>Specifies whether to output found pairs of sequences. -pri[ntinstanes] &lt;false&gt can be used as a quick way to count the number of pairs found. -pri[ntinstanes] &lt;true&gt is the default option</a></TD>
</TR>
 
</TABLE>
<p></p>
Example: <a href="./examples/example_2/example_2.txt">calculate (dn) repeats and write output to file</a>&nbsp;&nbsp;<a href="./examples/example_2/">Results</a>
<p><br></p>
<h3 id="seqtypes">Repeat types options</h3>
<p></p>
<TABLE CELLPADDING="0" CELLSPACING="5">
<TR ALIGN="LEFT">
<TD style="padding: 0 15px; vertical-align:top;"><a><i>-dn</i> </a></TD>
<TD style="padding: 0 15px;"><a> Search for direct non-complementary repeats </a></TD>
<TR ALIGN="LEFT">
</TR>
<TD style="padding: 0 15px; vertical-align:top;"><a><i>-dc</i></a></TD>
<TD style="padding: 0 15px;"><a>Search for direct complementary repeats</a></TD>
</TR>
<TR ALIGN="LEFT">
</TR>
<TD  style="padding: 0 15px; vertical-align:top;"><a><i>-in</i></a></TD>
<TD  style="padding: 0 15px;"><a>Search for inverse non-complementary repeats </a></TD>
</TR>
<TR ALIGN="LEFT">
<TD style="padding: 0 15px; vertical-align:top;"><a> <i>-ic</i></a></TD>
<TD style="padding: 0 15px;"><a>Search for inverse complementary repeats </a></TD>
</TR>
 
</TABLE>
<p></p>
If none of these options is specified the default is -dn.<br>
<p>Example: <a href="./examples/example_3/example_3.txt">calculate different types of repeats</a>&nbsp;&nbsp;<a href="./examples/example_3/">Results</a></p>
<p><br></p>
<h3 id="estimation">Probability estimation</h3>
StatRepeats uses probability estimation as explained in paper to estimate if a result is statistically significant. The following options can be specified: 
<p></p>
<TABLE CELLPADDING="0" CELLSPACING="5">
<TR ALIGN="LEFT">
<TD style="padding: 0 15px; vertical-align:top;"><a><i>-pv[alue] value</i> </a></TD>
<TD style="padding: 0 15px;"><a> Specify value of <i>p-value</i> parameter. Value is a number between 0 and 1. The default is 0.05 </a></TD>
<TR ALIGN="LEFT">
</TR>
<TD style="padding: 0 15px; vertical-align:top;"><a><i>-n[oprobability]</i></a></TD>
<TD style="padding: 0 15px;"><a>extract all repeat sequences without probability estimation</a></TD>
</TR>
 
</TABLE>
<p></p>
Example: <a href="./examples/example_4/example_4.txt">calculate different types of repeats</a>&nbsp;&nbsp;<a href="./examples/example_4/">Results</a>
<p><br></p>
<h3 id="alphabet">Option related to input sequence characteristics</h3>
Notes:
<ul>
<li>StatRepeats can handle an arbitrary alphabet. In the case of complementary repeat search letter complements must be specified. 
Input alphabet must include even number of characters (even - character and its complement)
<li>If input material includes characters not defined in the alphabet, then number of such character is reported, and 
each of these characters is assumed to be its own complement. 
<li>In options <i>aliphatic, sulphur, tiny, aromatic, hydrophobic, charged, positive, polar, acidic, small, </i>and <i>hydroxylic</i>  amino acids are groupped according to Esquivel et al:<i>Decoding the Building Blocks of Life from the Perspective of Quantum Information</i> in  "Advances in Quantum Mechanics", Paul Bracken (ed.), INTECH, Rijeka, pp: 641--669, (2013).
<br>
If some of this options is used options -protein must be specified for correct results. Multiple options from this group can be specified only if intersection of their AA groups is empty.
</ul>
<p></p>
Options: 
<p></p>
<TABLE CELLPADDING="0" CELLSPACING="5">
<TR ALIGN="LEFT">
<TD style="padding: 0 15px; vertical-align:top;"><a> <i>-dna</i>       </a></TD>
<TD style="padding: 0 15px;"><a> Specifies that the input sequence is DNA with the letters ACGT having the complements TGCA. This option is the default </a></TD>
<TR ALIGN="LEFT">
</TR>
<TD style="padding: 0 15px; vertical-align:top;"><a> <i>-rna</i></a></TD>
<TD style="padding: 0 15px;"><a> Specifies that the input sequence is RNA with the letters ACGU having the complements UGCA</a></TD>
</TR>
<TR ALIGN="LEFT">
</TR>
<TD  style="padding: 0 15px; width:250px; vertical-align:top;"><a><i>-compl[ement]&lt;filename&gt;</i></a></TD>
<TD  style="padding: 0 15px;"><a> Specifies the file with the letters and their complements. The file should be a text file with an even number of letters. The first two letters form a pair with the first letter having the second letter as a complement, then the second two letters form another pair, etc.
Statistical filtering is less precise for alphabets that do not consist of four or twenty letters in cases of complementary repeats and is not supported for non-complementary repeats.</a></TD>
</TR>
<TR ALIGN="LEFT">
<TD style="padding: 0 15px; vertical-align:top;"><a> <i>-ex[clude] len</i></a></TD>
<TD style="padding: 0 15px;"><a>Exclude every consecutive <i>len</i> or more letters N from repeat searching. The letter N is genomic sequences is used to specify an unknown nucleid acid. 
Large islands of N's typically lead to an explosion of found pairs. This option instructs StatRepeats to skip over any island of <i>len</i> or more consecutive N's. 
If -exclude 0 is specified than all N will be excluded.</a></TD>
</TR> 

<TR ALIGN="LEFT">
<TD style="padding: 0 15px; vertical-align:top;"><a> <i>-msr</i></a></TD>
<TD style="padding: 0 15px;"><a>Treat a multi-sequence FASTA file as a single large sequence. Both sequences in the output pair are characterized by the sequence name and intra-sequence position</a></TD>
</TR> 

<TR ALIGN="LEFT">
<TD style="padding: 0 15px; vertical-align:top;"><a><i>-max[gap] number</i></a></TD>
<TD style="padding: 0 15px;"><a>Specifies the maximal gap between two repeats in results. If max=0 then program is search for <i>tandem</i> repeats. If not, specified maximal gap is not set.</a></TD>
</TR> 


<TR ALIGN="LEFT">
<TD style="padding: 0 15px; vertical-align:top;"><a> <i>-protein</i></a></TD>
<TD style="padding: 0 15px;"><a>Specifies that the input sequence(s) are proteins. Proteins don't support extraction of complementary repeats as they don't have complements so this option just informs StatRepeats of the alphabet cardinality so that probabilities can be computed. Alphabet cardinality is 20, 21 or 22 depending on the input sequence. Different mappings are also provided when using proteins.</a></TD>
</TR>
<TR ALIGN="LEFT">
    <TD style="padding: 0 35px; vertical-align:top;"><a> <i>-acidic</i></a></TD>
    <TD style="padding: 0 15px;"><a> Specifies that in the input sequence amino acids N,Q will be mapped into simbol 8. Statistical estimation with new alphabet cardinality is enabled.</a></TD>
</TR>
<TR ALIGN="LEFT">
<TD style="padding: 0 35px; vertical-align:top;"><a> <i>-aliphatic</i></a></TD>
<TD style="padding: 0 15px;"><a> Specifies that in the input sequence amino acids I,V,L will be mapped into simbol 0. Statistical estimation with new alphabet cardinality is enabled.</a></TD>
</TR> 
<TR ALIGN="LEFT">
    <TD style="padding: 0 35px; vertical-align:top;"><a> <i>-aromatic</i></a></TD>
    <TD style="padding: 0 15px;"><a> Specifies that in the input sequence amino acids F,Y,W,H will be mapped into simbol 3. Statistical estimation with new alphabet cardinality is enabled.</a></TD>
</TR>
<TR ALIGN="LEFT">
    <TD style="padding: 0 35px; vertical-align:top;"><a> <i>-charged</i></a></TD>
    <TD style="padding: 0 15px;"><a> Specifies that in the input sequence amino acids D,E,H,K,R will be mapped into simbol 5. Statistical estimation with new alphabet cardinality is enabled.</a></TD>
</TR>
<TR ALIGN="LEFT">
    <TD style="padding: 0 35px; vertical-align:top;"><a> <i>-hydrophobic</i></a></TD>
    <TD style="padding: 0 15px;"><a> Specifies that in the input sequence amino acids A,C,T,K,H,W,Y,F,M,I,L,V will be mapped into simbol 4. Statistical estimation with new alphabet cardinality is enabled.</a></TD>
</TR>
<TR ALIGN="LEFT">
    <TD style="padding: 0 35px; vertical-align:top;"><a> <i>-hydroxylic</i></a></TD>
    <TD style="padding: 0 15px;"><a> Specifies that in the input sequence amino acids S,T will be mapped into simbol @. Statistical estimation with new alphabet cardinality is enabled.</a></TD>
</TR>
<TR ALIGN="LEFT">
    <TD style="padding: 0 35px; vertical-align:top;"><a> <i>-polar</i></a></TD>
    <TD style="padding: 0 15px;"><a> Specifies that in the input sequence amino acids N,Q,S,T,C,D,E,H,K,R,Y,W will be mapped into simbol 7. Statistical estimation with new alphabet cardinality is enabled.</a></TD>
</TR>
<TR ALIGN="LEFT">
    <TD style="padding: 0 35px; vertical-align:top;"><a> <i>-positive</i></a></TD>
    <TD style="padding: 0 15px;"><a> Specifies that in the input sequence amino acids H,K,R will be mapped into simbol 6. Statistical estimation with new alphabet cardinality is enabled.</a></TD>
</TR>


<TR ALIGN="LEFT">
    <TD style="padding: 0 35px; vertical-align:top;"><a> <i>-small</i></a></TD>
    <TD style="padding: 0 15px;"><a> Specifies that in the input sequence amino acids V,P,A,G,C,T,S,D,N will be mapped into simbol 9. Statistical estimation with new alphabet cardinality is enabled.</a></TD>
</TR>
<TR ALIGN="LEFT">
    <TD style="padding: 0 35px; vertical-align:top;"><a> <i>-sulphur</i></a></TD>
    <TD style="padding: 0 15px;"><a> Specifies that in the input sequence amino acids M,C will be mapped into simbol 1. Statistical estimation with new alphabet cardinality is enabled.</a></TD>
</TR>
<TR ALIGN="LEFT">
    <TD style="padding: 0 35px; vertical-align:top;"><a> <i>-tiny</i></a></TD>
    <TD style="padding: 0 15px;"><a> Specifies that in the input sequence amino acids A,G,C,S will be mapped into simbol 2. Statistical estimation with new alphabet cardinality is enabled.</a></TD>
</TR>






</TABLE>
<p></p>

Example:
<a href="./examples/example_5/example_5.txt">Options -dna, -rna, -compl</a>&nbsp;&nbsp;<a href="./examples/example_5/">Results</a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<a href="./examples/example_6/example_6.txt">Options -exclude, -proteins, -msr, -maxgap</a>&nbsp;&nbsp;<a href="./examples/example_6/">Results</a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<a href="./examples/example_7/example_7.txt">Options -proteins -aromatic -aliphatic -charged -tiny</a>&nbsp;&nbsp;<a href="./examples/example_7/">Results</a>
<p><br></p>    
<h3>Database Output</h3>
<p>
Results can be written directly either to DB2 or to other relational database with an ODBC driver
installed. A sample database schema can be found in <a href="./examples/example_8/">dbschema</a>. Order of SQL files execution is shown in <a href="./examples/example_8/example_8.txt">create dbschema</a>. Options are: </p>
<p></p>
<TABLE CELLPADDING="0" CELLSPACING="5">
<TR ALIGN="LEFT">
<TD style="padding: 0 15px; vertical-align:top;"><a><i>-dat[abase] &lt;ODBC connection string&gt;</i> </a></TD>
<TD style="padding: 0 15px;"><a> Specify ODBC connection string used to connect to database</a></TD>
<TR ALIGN="LEFT">
</TR>
<TD style="padding: 0 15px; vertical-align:top;"><a><i>-db2[output] &lt;connection string&gt;</i></a></TD>
<TD style="padding: 0 15px;"><a>Specify connection string (not necessary ODBC)  used to connect to DB2 database</a></TD>
</TR>
<TR ALIGN="LEFT">
<TD style="padding: 0 15px; vertical-align:top;"><a><i>-com[mitcount] number</i> </a></TD>
<TD style="padding: 0 15px;"><a> Specifies how many inserts are done in one transaction. Default is 0 (only one commit is done at the end of the program</a></TD>
<TR ALIGN="LEFT">
</TR>
<TD style="padding: 0 15px; vertical-align:top;"><a><i>-log[ging] true/false</i></a></TD>
<TD style="padding: 0 15px;"><a>Used only with DB2 row oriented database. Default is true. If false, than <i>alter table ... activate not logged initially</i> is executed in the beginning to the transaction 
on fragment and match tables</a></TD>
</TR> 
</TABLE>
<p></p>
Example: <a href="./examples/example_8/example_8.txt">load results directly to relational database</a>&nbsp;&nbsp;<a href="./examples/example_8/">Database schema</a>
<p><br></p>

<h2 id="download">Download</h2>
<p>Source code of StatRepeats can be download from site <a href="http://bioinfo.matf.bg.ac.rs/home/downloads.wafl?cat=Software\&project=StatRepeats">	
Bioinformatics Research Group at Faculty of Mathematics, University of Belgrade.</a> 
<br>
Distribution directory includes precompiled version for 64 bit Linux (Ubuntu), 64-bit Windows and license (StatRepeats and Divsufsort) files. Two verson o program are available: with (StatRepeats) or without (StatRepeatsNoDB) interface to relational database. Suffix in program name denotes curent version and release (for example, StatRepeats.v1.r1).  


<h2 id="setup">Setup Instructions</h2>

<h3 id="linux">Linux</h3>
<p>To compile <i>StatRepeats</i> execute the following command in the directory where you unpacked the archive:</p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;<i> ./StatRepeats.compile.sh</i> for program with interface to relational database </p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;<i> ./StatRepeatsNoDB.compile.sh</i> for program without interface to relational database </p>
<p>These commands produce the <i>StatRepeats.version.release</i> or <i>StatRepeatsNoDB.version.release</i> binaries. In order for the above command to succeed you will need gcc 4.8.4 or later and libodbc (for version with relational database interface).</p>

    <p> In case you are using Debian or a derived distribution the typical
way to install ODBC support which is needed to compile StatRepeats is by
running the following commands:</p>

<p>&nbsp;&nbsp;&nbsp;&nbsp;<i>sudo apt-get install odbcinst</i></p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;<i>sudo apt-get install unixodbc-dev</i></p>
<p>For other distributions please check your package manager documentation or build from source.</p>

<h3 id="windows">Windows</h3>
<p>StatRepeats executable version (<i>StatRepeats.version.release.exe</i>) is provided in this archive. You will need to install the Visual C++ 2015 runtime in order to run it. It can be downloaded from 
<a href="https://www.microsoft.com/en-us/download/details.aspx?id=48145">Microsoft site</a>.</p>
<p> To compile from source you will need the free Visual Studio 2015 Community Edition. Open StatRepeats.sln which is included in the archive and build the solution.</p>
</body>
</html>

